### PR TITLE
refactor `board`

### PR DESCRIPTION
**Pull request by Dillon-Roller**

### CODE REVIEW

File: src/board.rs [hunk 1]
```diff
@@ -56,6 +56,10 @@ impl Board {
         self.squares[pos.y][pos.x]
     }
 
+    pub fn get_piece_mut(&mut self, pos: &Position) -> &mut Option<Piece> {
```
Comment (DanielAlbl) [inline src/board.rs:59]: Do we want this to be public?

File: src/board.rs [hunk 2]
```diff
@@ -84,18 +88,18 @@ impl Board {
...
-        let mut piece = self.get_piece(&m.from);
-
-        if let Some(Piece::Pawn(player)) = piece {
-            if (player == Player::White && m.to.y == 7) || (player == Player::Black && m.to.y == 0)
-            {
-                // TODO: always promote to queen for now, need to handle this eventually
-                piece = Some(Piece::Queen(player));
+        self.is_move_valid(m)?;
+
+        if let Some(mut piece) = self.get_piece_mut(&m.from).take() {
```
Comment (DanielAlbl) [inline src/board.rs:93]: So if we know it is a valid move at this point, would it be cleaner to `let piece ... .unwrap()`?

Comment (Dillon-Roller) [reply src/board.rs:93]: No because I want to borrow the piece mutably and then change it

Comment (Dillon-Roller) [reply src/board.rs:93]: Note the `mut piece `

Comment (DanielAlbl) [reply src/board.rs:93]: Yeah I mean `let mut piece ...`

Comment (DanielAlbl) [reply src/board.rs:93]: `let mut piece = self.get_piece_mut(&m.from).take().unwrap();`

File: src/board.rs [hunk 3]
```diff
@@ -115,11 +116,8 @@ impl Board {
             //check for double move
             to += v;
             if self.get_piece(&to).is_none() {
-                let can_double_move = match self.get_piece(&from).unwrap().get_player() {
-                    Player::White => from.y == 1,
-                    Player::Black => from.y == 6,
-                };
-                if can_double_move {
+                let player = self.get_piece(&from).unwrap().get_player();
+                if [from.y == 1, from.y == 6][player as usize] {
```
Comment (DanielAlbl) [inline src/board.rs:120]: Tiger Woods ass code golf.

File: src/board.rs [hunk 4]
```diff
@@ -72,38 +76,35 @@ impl Board {
...
 
-        if let Some(Piece::Pawn(player)) = piece {
-            if (player == Player::White && m.to.y == 7) || (player == Player::Black && m.to.y == 0)
-            {
-                // TODO: always promote to queen for now, need to handle this eventually
-                piece = Some(Piece::Queen(player));
+        if let Some(mut piece) = self.get_piece_mut(&mv.from).take() {
+            if let Piece::Pawn(player) = piece {
+                if (player == Player::White && mv.to.y == BOARD_SIZE - 1)
+                    || (player == Player::Black && mv.to.y == 0)
```
Comment (DanielAlbl) [inline src/board.rs:96]: Can we use that code golf technique here as well?

File: src/board.rs [hunk 5]
```diff
@@ -52,8 +52,12 @@ impl Board {
         ]
     }
 
-    pub fn get_piece(&self, pos: &Position) -> Option<Piece> {
-        self.squares[pos.y][pos.x]
+    pub fn get_piece(&self, pos: &Position) -> &Option<Piece> {
```
Comment (DanielAlbl) [inline src/board.rs:55]: What's the argument for returning `&Option` here?

Comment (Dillon-Roller) [reply src/board.rs:55]: probably just borrow until we need to clone

File: src/board.rs (hunk 4 above)
Comment (Dillon-Roller) [reply src/board.rs:96]: I almost kind of don't want that code golf now lol

File: src/board.rs (hunk 5 above)
Comment (DanielAlbl) [reply src/board.rs:55]: Isn't `Option<Piece>` smaller than a reference?

Comment (Dillon-Roller) [reply src/board.rs:55]: so it does matter?

Comment (DanielAlbl) [reply src/board.rs:55]: It either does it which case we should copy to save memory, or it doesn't in which case we should copy to make the code easier to work with.

File: src/pieces.rs [hunk 6]
```diff
@@ -46,6 +46,16 @@ pub enum Player {
     Black,
 }
 
+impl Not for Player {
```
Comment (DanielAlbl) [inline src/pieces.rs:49]: Oh damn that is clean

File: src/board.rs [hunk 7]
```diff
@@ -52,119 +53,127 @@ impl Board {
...
-            _ => {
-                for &v in piece.get_vectors() {
-                    let mut to = from + v;
-                    while Self::is_in_bounds(&to).is_ok() {
-                        if let Some(piece) = self.get_piece(&to) {
-                            if piece.get_player() != self.player {
-                                self.moves.insert(Move { from, to });
+    fn add_moves_for_piece(&mut self, from: Position) {
+        if let Some(piece) = self.get_piece(from) {
+            if piece.get_player() == self.player {
```
Comment (DanielAlbl) [inline src/board.rs:157]: How do you feel about `if piece.get_player() != self.player { return }`?

File: src/board.rs [hunk 8]
```diff
@@ -3,6 +3,7 @@ use crate::game::{ChessError, ChessResult};
 use crate::moves::Move;
 use crate::pieces::{Piece, Player, Position};
 use std::collections::HashSet;
+use std::mem;
```
Comment (DanielAlbl) [inline src/board.rs:6]: Do we need this?

File: src/board.rs [hunk 9]
```diff
@@ -52,119 +52,127 @@ impl Board {
...
-            if (player == Player::White && m.to.y == 7) || (player == Player::Black && m.to.y == 0)
-            {
-                // TODO: always promote to queen for now, need to handle this eventually
-                piece = Some(Piece::Queen(player));
-            }
+    pub fn move_piece(&mut self, mv: &Move) -> ChessResult<()> {
+        self.is_move_valid(mv)?;
+
+        let mut piece = self.take_piece(&mv.from);
+        if self.can_promote_piece(&mv.from) {
```
Comment (DanielAlbl) [inline src/board.rs:99]: `to` not `from` right?

File: src/chess_widget.rs [hunk 10]
```diff
@@ -103,7 +103,7 @@ impl ChessWidget {
             .map(|file_name| create_image(file_name, ctx, PIECE_SIZE, ImageFormat::RgbaSeparate))
     }
 
-    fn get_image_file(&self, piece: Piece) -> &PietImage {
+    fn get_image_file(&self, piece: &Piece) -> &PietImage {
```
Comment (DanielAlbl) [inline src/chess_widget.rs:106]: Not necessary

**Review by DanielAlbl — APPROVED** [review]

**Pull request merged**